src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-core.c

```diff
@@ -898,7 +899,28 @@ void xclmgmt_mailbox_srv(void *arg, void *data, size_t len,
 			mgmt_err(lro, "peer request dropped, wrong size\n");
 			break;
 		}
-		ret = xocl_xclbin_download(lro, xclbin);
+
+		/*
+		 * User may transfer a fake xclbin which doesn't have bitstream
+		 * In this case, 'config_xclbin_change' has to be set, and we
+		 * will go to fetch the real xclbin.
+		 * Note:
+		 * 1. it is up to the admin to put authentificated xclbins at
+		 *    predefined location
+		 * 2. the fetched xclbin buf need to be freed here after use
+		 */
+		if (fetch)
+			xclbin = xclmgmt_xclbin_fetch(lro, xclbin);
```

Comment:
I suggest to avoid alloc memory inside and free outside. Maybe it is better to just put xocl_xclbin_download() and vfree(xclbin) into xclgmmt_xclbin_fetch()?

Reply:
changed